File: brorb/cpg.py
"""
Brainstem respiratory CPG model.

Activity-based model after Molkov et al. (2017) / Rubin et al. (2011).
Five neural populations with 10 coupled ODEs producing a three-phase
respiratory rhythm (inspiration → post-inspiration → late expiration).

Populations:
  1. pre-I/I   (excitatory, pre-BötC) — INaP bursting
  2. early-I   (inhibitory, pre-BötC) — adaptation
  3. post-I    (inhibitory, BötC)     — adaptation
  4. aug-E     (inhibitory, BötC)     — adaptation
  5. late-E    (excitatory, RTN/pFRG) — INaP bursting

State vector [10]:
  V1, V2, V3, V4, V5, hNaP1, hNaP5, mAD2, mAD3, mAD4

Uses current-based synaptic formulation. The key to oscillation:
- Pre-I/I bursts via INaP, terminates when h1 drops (slow inactivation)
- Early-I follows pre-I/I, self-terminates via adaptation (m_AD2 rises)
- As inspiration dies, post-I escapes inhibition and becomes active
- Post-I self-terminates via m_AD3, letting aug-E take over
- Aug-E self-terminates via m_AD4, and pre-I/I restarts (h1 recovered)
"""

import numpy as np


def sigmoid(V, k=10.0, V_half=0.0):
    """Firing rate sigmoid: f(V) = 1 / (1 + exp(-k*(V - V_half)))"""
    x = k * (V - V_half)
    return 1.0 / (1.0 + np.exp(-np.clip(x, -50, 50)))


def default_params(target_bpm=4.0):
    """Return parameter dict tuned for the given breathing rate."""
    period = 60.0 / target_bpm  # ~15s at 4 bpm

    p = {}

    # ── Membrane time constant ────────────────────────────────────
    p["tau_m"] = 0.02

    # ── Tonic drives ──────────────────────────────────────────────
    # Crucial: tonic drives determine which populations can self-activate.
    # Post-I needs enough drive to turn on when disinhibited.
    p["d1"] = 0.0     # pre-I/I: relies entirely on INaP for excitation
    p["d2"] = -0.3    # early-I: requires pre-I/I drive
    p["d3"] = 0.3     # post-I: self-activating when disinhibited
    p["d4"] = 0.15    # aug-E: can self-activate when disinhibited
    p["d5"] = -0.5    # late-E: normally quiescent

    # ── INaP (pops 1, 5) ─────────────────────────────────────────
    # The burst condition: when h is high, g_NaP * h > threshold,
    # creating regenerative excitation. As h drops, the neuron shuts off.
    # With V equation: dV/dt = (-V + g_NaP*σ(V)*h + syn + drive) / tau_m
    # At the V=0 fixed point, σ(0, k=8, Vh=0) = 0.5
    # For burst: g_NaP * 0.5 * h_high > 0 → burst starts
    # Burst terminates when g_NaP * σ(V) * h_low + drive + syn < V_eq
    p["g_NaP1"] = 3.0
    p["g_NaP5"] = 2.5

    # h_NaP slow inactivation
    # With steep slope and low Vh, h_inf is near 0 for any V > 0,
    # so h decays monotonically during the entire burst.
    p["k_hNaP"] = -10.0   # steep inactivation
    p["Vh_hNaP"] = -0.2   # h_inf ≈ 0 whenever V > 0 (always decaying during burst)
    p["tau_hNaP"] = period * 0.6   # inspiration duration scaling

    # ── Adaptation (pops 2, 3, 4) ─────────────────────────────────
    p["g_AD2"] = 4.0     # strong adaptation to ensure self-termination
    p["g_AD3"] = 3.5
    p["g_AD4"] = 3.0

    p["k_mAD"] = 6.0     # activation slope
    p["Vh_mAD"] = 0.1    # activates when V > 0.1
    p["tau_AD2"] = period * 0.47  # early-I adapts within inspiration
    p["tau_AD3"] = period * 0.73  # post-I holds through post-I phase
    p["tau_AD4"] = period * 0.93  # aug-E holds through E2 phase

    # ── Recovery time constants ───────────────────────────────────
    # When a pop is inactive, its slow variable recovers.
    # h_NaP recovers (increases) — tau is the same (governed by h_inf)
    # m_AD recovers (decreases) — tau could be different for recovery
    # We handle this asymmetry by making tau_AD voltage-dependent
    # (faster decay when V is low). For simplicity, use same tau but
    # ensure h_inf / m_AD_inf provide strong gradient.

    # ── Synaptic weights ──────────────────────────────────────────
    # FROM pre-I/I (pop 1) — excitatory
    p["w_21"] = 1.2     # → early-I

    # FROM early-I (pop 2) — inhibitory
    p["w_32"] = -2.0    # → post-I
    p["w_42"] = -1.5    # → aug-E
    p["w_52"] = -1.0    # → late-E

    # FROM post-I (pop 3) — inhibitory
    p["w_13"] = -2.5    # → pre-I/I (strong! must kill INaP burst)
    p["w_23"] = -1.5    # → early-I
    p["w_43"] = -2.0    # → aug-E
    p["w_53"] = -0.8    # → late-E

    # FROM aug-E (pop 4) — inhibitory
    p["w_14"] = -1.5    # → pre-I/I
    p["w_24"] = -1.0    # → early-I
    p["w_34"] = -2.5    # → post-I (strong! must kill post-I)
    p["w_54"] = -0.5    # → late-E

    # FROM late-E (pop 5) — excitatory
    p["w_15"] = 0.5     # → pre-I/I

    # ── Firing rate sigmoid ───────────────────────────────────────
    p["k_f"] = 8.0
    p["Vh_f"] = 0.0      # threshold at 0

    return p


class CPGState:
    """Holds the 10 state variables as a numpy array."""

    V1, V2, V3, V4, V5 = 0, 1, 2, 3, 4
    H1, H5 = 5, 6
    M2, M3, M4 = 7, 8, 9
    N = 10

    def __init__(self, y=None):
        if y is not None:
            self.y = np.array(y, dtype=np.float64)
        else:
            # Start in late expiration (about to inspire)
            self.y = np.array([
                -0.5,   # V1: pre-I/I (inactive)
                -0.5,   # V2: early-I (inactive)
                -0.5,   # V3: post-I (inactive)
                 0.3,   # V4: aug-E (active, about to exhaust)
                -0.5,   # V5: late-E (inactive)
                 0.95,  # hNaP1: fully de-inactivated (ready to burst)
                 0.95,  # hNaP5: fully de-inactivated
                 0.02,  # mAD2: low
                 0.02,  # mAD3: low
                 0.6,   # mAD4: high (aug-E has been active, about to adapt)
            ], dtype=np.float64)


def cpg_derivatives(y, p, external_drives=None):
    """
    Compute dy/dt for the 10-variable CPG model.

    Voltage equation:
      tau_m * dV/dt = -V + I_NaP(V, h) + I_syn(f_j) + I_drive - g_AD * m_AD

    The -V term provides a unitary leak. All currents are additive.
    INaP is the only intrinsic excitatory current (pops 1, 5).
    Adaptation is the only intrinsic inhibitory current (pops 2, 3, 4).
    """
    ext = external_drives or {}
    dydt = np.zeros(10, dtype=np.float64)

    V1, V2, V3, V4, V5 = y[0], y[1], y[2], y[3], y[4]
    h1, h5 = y[5], y[6]
    m2, m3, m4 = y[7], y[8], y[9]

    # Firing rates
    kf, vf = p["k_f"], p["Vh_f"]
    f1 = sigmoid(V1, kf, vf)
    f2 = sigmoid(V2, kf, vf)
    f3 = sigmoid(V3, kf, vf)
    f4 = sigmoid(V4, kf, vf)
    f5 = sigmoid(V5, kf, vf)

    # INaP: I = g * sigmoid(V) * h  (positive feedback, gated by slow h)
    I_NaP1 = p["g_NaP1"] * sigmoid(V1, 10.0, -0.1) * h1
    I_NaP5 = p["g_NaP5"] * sigmoid(V5, 10.0, -0.1) * h5

    # Synaptic currents (current-based)
    I_syn1 = p["w_13"]*f3 + p["w_14"]*f4 + p["w_15"]*f5
    I_syn2 = p["w_21"]*f1 + p["w_23"]*f3 + p["w_24"]*f4
    I_syn3 = p["w_32"]*f2 + p["w_34"]*f4
    I_syn4 = p["w_42"]*f2 + p["w_43"]*f3
    I_syn5 = p["w_52"]*f2 + p["w_53"]*f3 + p["w_54"]*f4

    # External drives
    ext1 = p["d1"] + ext.get("drive_1", 0.0)
    ext2 = p["d2"] + ext.get("drive_2", 0.0)
    ext3 = p["d3"] + ext.get("drive_3", 0.0)
    ext4 = p["d4"] + ext.get("drive_4", 0.0)
    ext5 = p["d5"] + ext.get("drive_5", 0.0)

    # Voltage dynamics
    tau = p["tau_m"]
    dydt[0] = (-V1 + I_NaP1           + I_syn1 + ext1) / tau  # pre-I/I
    dydt[1] = (-V2 - p["g_AD2"] * m2  + I_syn2 + ext2) / tau  # early-I
    dydt[2] = (-V3 - p["g_AD3"] * m3  + I_syn3 + ext3) / tau  # post-I
    dydt[3] = (-V4 - p["g_AD4"] * m4  + I_syn4 + ext4) / tau  # aug-E
    dydt[4] = (-V5 + I_NaP5           + I_syn5 + ext5) / tau  # late-E

    # h_NaP inactivation: h_inf decreases with V (k < 0)
    h_inf1 = sigmoid(V1, p["k_hNaP"], p["Vh_hNaP"])
    h_inf5 = sigmoid(V5, p["k_hNaP"], p["Vh_hNaP"])
    dydt[5] = (h_inf1 - h1) / p["tau_hNaP"]
    dydt[6] = (h_inf5 - h5) / p["tau_hNaP"]

    # m_AD adaptation: m_inf increases with V (k > 0)
    mAD_inf2 = sigmoid(V2, p["k_mAD"], p["Vh_mAD"])
    mAD_inf3 = sigmoid(V3, p["k_mAD"], p["Vh_mAD"])
    mAD_inf4 = sigmoid(V4, p["k_mAD"], p["Vh_mAD"])
    dydt[7] = (mAD_inf2 - m2) / p["tau_AD2"]
    dydt[8] = (mAD_inf3 - m3) / p["tau_AD3"]
    dydt[9] = (mAD_inf4 - m4) / p["tau_AD4"]

    return dydt


def rk4_step(y, dt, p, external_drives=None):
    """Single RK4 integration step."""
    k1 = cpg_derivatives(y, p, external_drives)
    k2 = cpg_derivatives(y + 0.5 * dt * k1, p, external_drives)
    k3 = cpg_derivatives(y + 0.5 * dt * k2, p, external_drives)
    k4 = cpg_derivatives(y + dt * k3, p, external_drives)
    return y + (dt / 6.0) * (k1 + 2*k2 + 2*k3 + k4)

File: sim/cpg.js
/**
 * Brainstem respiratory CPG model.
 *
 * Activity-based model after Molkov et al. (2017) / Rubin et al. (2011).
 * Five neural populations with 10 coupled ODEs producing a three-phase
 * respiratory rhythm (inspiration -> post-inspiration -> late expiration).
 *
 * Populations:
 *   1. pre-I/I   (excitatory, pre-BotC) — INaP bursting
 *   2. early-I   (inhibitory, pre-BotC) — adaptation
 *   3. post-I    (inhibitory, BotC)     — adaptation
 *   4. aug-E     (inhibitory, BotC)     — adaptation
 *   5. late-E    (excitatory, RTN/pFRG) — INaP bursting
 *
 * State vector [10]:
 *   V1, V2, V3, V4, V5, hNaP1, hNaP5, mAD2, mAD3, mAD4
 *
 * Uses current-based synaptic formulation.
 */

const N = 10;

// State indices
const V1 = 0, V2 = 1, V3 = 2, V4 = 3, V5 = 4;
const H1 = 5, H5 = 6;
const M2 = 7, M3 = 8, M4 = 9;

export function sigmoid(V, k = 10.0, Vhalf = 0.0) {
    const x = Math.max(-50, Math.min(50, k * (V - Vhalf)));
    return 1.0 / (1.0 + Math.exp(-x));
}

/**
 * Default CPG parameters with FIXED biophysical timescales.
 *
 * Time constants are intrinsic membrane properties (Rubin et al. 2011,
 * J Comput Neurosci 30:607-632) — they do NOT change with breathing rate.
 * Rate modulation works through tonic drives d1–d5 (Molkov et al. 2017).
 */
export function defaultParams() {
    return {
        tau_m: 0.02,          // membrane time constant, 20ms (fast)

        // Tonic drives — baseline for ~4 BPM eupnea
        d1: 0.0, d2: -0.3, d3: 0.3, d4: 0.15, d5: -0.5,

        // INaP inactivation — FIXED biophysical (pops 1, 5)
        g_NaP1: 3.0, g_NaP5: 2.5,
        k_hNaP: -10.0, Vh_hNaP: -0.2,
        tau_hNaP: 9.0,        // INaP inactivation, ~9s

        // Adaptation — FIXED biophysical (pops 2, 3, 4)
        g_AD2: 4.0, g_AD3: 3.5, g_AD4: 3.0,
        k_mAD: 6.0, Vh_mAD: 0.1,
        tau_AD2: 7.0,         // early-I adaptation, ~7s
        tau_AD3: 11.0,        // post-I adaptation, ~11s
        tau_AD4: 14.0,        // aug-E adaptation, ~14s

        // Synaptic weights
        w_21: 1.2,                          // pre-I/I -> early-I (exc)
        w_32: -2.0, w_42: -1.5, w_52: -1.0, // early-I -> (inh)
        w_13: -2.5, w_23: -1.5, w_43: -2.0, w_53: -0.8, // post-I -> (inh)
        w_14: -1.5, w_24: -1.0, w_34: -2.5, w_54: -0.5, // aug-E -> (inh)
        w_15: 0.5,                          // late-E -> pre-I/I (exc)

        // Firing rate sigmoid
        k_f: 8.0, Vh_f: 0.0,
    };
}

/**
 * Baseline tonic drives (for drive interpolation).
 * Separated from defaultParams so simulation can interpolate toward targets.
 */
export const BASE_DRIVES = { d1: 0.0, d2: -0.3, d3: 0.3, d4: 0.15, d5: -0.5 };

/**
 * Compute drive offsets for a target breathing rate.
 *
 * Per Molkov et al. (2017): rate is controlled by adjusting external
 * excitatory drives, not timescales. Increasing d1 (pre-I/I) strengthens
 * inspiration; decreasing d3 (post-I) shortens the post-inspiratory
 * pause; increasing d5 (late-E) recruits active expiration at high rates.
 *
 * @param {number} targetBpm - desired breaths per minute
 * @returns {{ d1_offset: number, d3_offset: number, d5_offset: number }}
 */
export function driveProfile(targetBpm) {
    const ratio = targetBpm / 4.0;
    return {
        d1_offset: (ratio - 1.0) * 0.4,
        d3_offset: (1.0 - ratio) * 0.2,
        d5_offset: Math.max(0, ratio - 2.0) * 0.3,
    };
}

export class CPGState {
    constructor(y = null) {
        if (y) {
            this.y = new Float64Array(y);
        } else {
            this.y = new Float64Array([
                -0.5,   // V1: pre-I/I (inactive)
                -0.5,   // V2: early-I (inactive)
                -0.5,   // V3: post-I (inactive)
                 0.3,   // V4: aug-E (active, about to exhaust)
                -0.5,   // V5: late-E (inactive)
                 0.95,  // hNaP1: fully de-inactivated
                 0.95,  // hNaP5: fully de-inactivated
                 0.02,  // mAD2: low
                 0.02,  // mAD3: low
                 0.6,   // mAD4: high (aug-E about to adapt)
            ]);
        }
    }
}

// Element-wise: out[i] = a[i] + s * b[i]
function addScaled(out, a, s, b) {
    for (let i = 0; i < N; i++) out[i] = a[i] + s * b[i];
    return out;
}

export function cpgDerivatives(y, p, ext, dydt) {
    const v1 = y[V1], v2 = y[V2], v3 = y[V3], v4 = y[V4], v5 = y[V5];
    const h1 = y[H1], h5 = y[H5];
    const m2 = y[M2], m3 = y[M3], m4 = y[M4];

    // Firing rates
    const kf = p.k_f, vf = p.Vh_f;
    const f1 = sigmoid(v1, kf, vf);
    const f2 = sigmoid(v2, kf, vf);
    const f3 = sigmoid(v3, kf, vf);
    const f4 = sigmoid(v4, kf, vf);
    const f5 = sigmoid(v5, kf, vf);

    // INaP: I = g * sigmoid(V) * h
    const iNaP1 = p.g_NaP1 * sigmoid(v1, 10.0, -0.1) * h1;
    const iNaP5 = p.g_NaP5 * sigmoid(v5, 10.0, -0.1) * h5;

    // Synaptic currents (current-based)
    const iSyn1 = p.w_13 * f3 + p.w_14 * f4 + p.w_15 * f5;
    const iSyn2 = p.w_21 * f1 + p.w_23 * f3 + p.w_24 * f4;
    const iSyn3 = p.w_32 * f2 + p.w_34 * f4;
    const iSyn4 = p.w_42 * f2 + p.w_43 * f3;
    const iSyn5 = p.w_52 * f2 + p.w_53 * f3 + p.w_54 * f4;

    // External drives
    const ext1 = p.d1 + (ext.drive_1 ?? 0);
    const ext2 = p.d2 + (ext.drive_2 ?? 0);
    const ext3 = p.d3 + (ext.drive_3 ?? 0);
    const ext4 = p.d4 + (ext.drive_4 ?? 0);
    const ext5 = p.d5 + (ext.drive_5 ?? 0);

    // Voltage dynamics
    const tau = p.tau_m;
    dydt[V1] = (-v1 + iNaP1           + iSyn1 + ext1) / tau;
    dydt[V2] = (-v2 - p.g_AD2 * m2    + iSyn2 + ext2) / tau;
    dydt[V3] = (-v3 - p.g_AD3 * m3    + iSyn3 + ext3) / tau;
    dydt[V4] = (-v4 - p.g_AD4 * m4    + iSyn4 + ext4) / tau;
    dydt[V5] = (-v5 + iNaP5           + iSyn5 + ext5) / tau;

    // h_NaP inactivation
    const hInf1 = sigmoid(v1, p.k_hNaP, p.Vh_hNaP);
    const hInf5 = sigmoid(v5, p.k_hNaP, p.Vh_hNaP);
    dydt[H1] = (hInf1 - h1) / p.tau_hNaP;
    dydt[H5] = (hInf5 - h5) / p.tau_hNaP;

    // m_AD adaptation
    const mInf2 = sigmoid(v2, p.k_mAD, p.Vh_mAD);
    const mInf3 = sigmoid(v3, p.k_mAD, p.Vh_mAD);
    const mInf4 = sigmoid(v4, p.k_mAD, p.Vh_mAD);
    dydt[M2] = (mInf2 - m2) / p.tau_AD2;
    dydt[M3] = (mInf3 - m3) / p.tau_AD3;
    dydt[M4] = (mInf4 - m4) / p.tau_AD4;

    return dydt;
}

// Pre-allocated scratch arrays for RK4
const _k1 = new Float64Array(N);
const _k2 = new Float64Array(N);
const _k3 = new Float64Array(N);
const _k4 = new Float64Array(N);
const _tmp = new Float64Array(N);

const _emptyExt = {};

export function rk4Step(y, dt, p, ext) {
    const e = ext ?? _emptyExt;
    cpgDerivatives(y, p, e, _k1);

    addScaled(_tmp, y, 0.5 * dt, _k1);
    cpgDerivatives(_tmp, p, e, _k2);

    addScaled(_tmp, y, 0.5 * dt, _k2);
    cpgDerivatives(_tmp, p, e, _k3);

    addScaled(_tmp, y, dt, _k3);
    cpgDerivatives(_tmp, p, e, _k4);

    // y_new = y + (dt/6) * (k1 + 2*k2 + 2*k3 + k4)
    const s = dt / 6.0;
    for (let i = 0; i < N; i++) {
        y[i] += s * (_k1[i] + 2 * _k2[i] + 2 * _k3[i] + _k4[i]);
    }
    return y;
}
